### DISCUSSION TITLE
How to use next-auth with Next 13 RSC's

### DISCUSSION THREAD

**Discussion by niels-bosman** [nextauthjs/next-auth#8650] [Help] [answered]
### Question 💬

How do I use next-auth for my authentication while keeping my react server components?

I now have the following `app/layout.tsx`:

```tsx
'use client';
import './globals.css';
import { ReactNode, Suspense } from 'react';
import Analytics from '@/components/Analytics';
import { Toaster } from 'react-hot-toast';
import { SessionProvider } from 'next-auth/react';
import { getServerSession } from 'next-auth';
import { authOptions } from '@/app/api/auth/[...nextauth]/route';

type RootLayoutProps = {
  children: ReactNode;
};

const RootLayout = async ({ children }: RootLayoutProps) => {
  const session = await getServerSession(authOptions);

  return (
    <SessionProvider session={session}>
      <html lang="en">
        <body className={font.className}>
          <Toaster />
          <Suspense>
            <Analytics />
          </Suspense>
          {children}
        </body>
      </html>
    </SessionProvider>
  );
};

export default RootLayout;
```

If I remove 'use client';, I get the following error:

```
Error: React Context is unavailable in Server Components
```

### How to reproduce ☕️

A `app/layout.tsx` with the content from my question.

A `app/api/auth/[...nextauth]/route.ts` with the following content:

```ts
import NextAuth, { AuthOptions } from 'next-auth';
import EmailProvider from 'next-auth/providers/email';

export const authOptions: AuthOptions = {
  providers: [
    EmailProvider({
      server: process.env.EMAIL_SERVER,
      from: process.env.EMAIL_FROM,
    }),
  ],
  session: { strategy: 'jwt' },
};

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```

nodemailer and next-auth packages installed.

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ChristofferWolgast** [comment]
Just put the SessionProvider in its own client component marked with 'use client' and import it in your layout.

**Comment by niels-bosman** [comment]
@ChristofferWolgast This does not answer my question _**(How do I use next-auth for my authentication while keeping my react server components?)**_, this makes it so every child component is a client component (which is not desired).

**Comment by J4v4Scr1pt** [comment]
Im also really curious of this, because as it seems according to [this ](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#using-context-providers)that @ChristofferWolgast is correct.

But at the same time [this](https://nextjs.org/docs/app/building-your-application/rendering/client-components#using-client-components-in-nextjs) states that: _"This means that by defining a "use client" in a file, all other modules imported into it, including child components, are considered part of the client bundle - and will be rendered by React on the client."_

Which is really confusing tbh, maybe some author could clearify this?

Edit, to add some more confusion 😅:
https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props

_With this approach, ClientComponent and ServerComponent are decoupled and can be rendered independently. In this case, the child <ServerComponent> can be rendered on the server, well before <ClientComponent> is rendered on the client._

**Comment by AR-Project** [comment]
I've also dealing this problem for past week, and I literally just found out the solution to use session without marking any component into client component.

Based on [this article on migrating data fetching](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#server-side-rendering-getserversideprops), I just put ```getServerSession(authOptions);``` on different async function, then await it inside my component. So its like

```typescript
// getSession.ts
import { getServerSession } from "next-auth/next";
import authOption from "./authOption";

export default async function getSession() {
  const session = await getServerSession(authOption);
  return session;
}
````

And on the react Component I just import the function like usual, without converting into client component.

```javascript
// MyComponent.tsx - without any use of 'use client'
import getSession from "./getSession";

export default async function MyComponent() {
  const session = await getSession();
  
  if (session) {
    return (<div>You're not logged In</div>)
  }

  return (
    <div>Welcome {`${session.user.name}}    </div>
  );
```
Please not that I just a newbie Developer who still learn JS, and this is still naive solution. I still believe there must be better solution somewhere someplace. 

Hope this helps you.

**Reply by niels-bosman** [reply]
I think this solution does fix the issue mentioned, but this does make it so the `useSession()` hook probably won't work in client components.

**Reply by AR-Project** [reply]
Yeah, you're right.

My first idea, (ofc its  naive solution ) is just wrap my function inside `useEffect()` but it gave me an error. 😅

> How do I use next-auth for my authentication while keeping my React server components?

With my limited knowledge, I think your answer to your question boils down to which way you want a particular page / path to render.

Right now, I've used both solutions from my answer above. I use `getServerSession` inside wrapper function if I want to render a page as a server component. And if I still want to use `useSession` I use @ChristofferWolgast 's solution below.

Personally, I also dislike flagging the root layout with `'use client'` ( because of [this](https://nextjs.org/docs/app/building-your-application/rendering/client-components#using-client-components-in-nextjs) ). So, if I need to use the `useSession` hook, I just implement the `SessionProvider` inside my sub-path `layout.tsx` where I want to implement it, instead of putting it on root `layout.tsx`.

**Comment by ChristofferWolgast** [comment, marked as answer]
Having a Provider (that has to have 'use client') that wraps your app does not mean that everything in it will be client components.
The proper solution is to have a provider in your layout, then use `getServerSession` in your RSC and `useSession` in client components.

```TypeScript
'use client'

import { SessionProvider, type SessionProviderProps } from 'next-auth/react'

export default function Provider(props: SessionProviderProps) {
  return (
    <SessionProvider session={props.session}>{props.children}</SessionProvider>
  )
}
```

**Comment by J4v4Scr1pt** [comment]
Just to add more context. Great video explaning this! :) 
https://youtu.be/3Dw6D_WuzSE?t=632
